src/Microsoft.Graph.Cli.Core/IO/AuthenticationCacheUtility.cs

```diff
@@ -0,0 +1,63 @@
+using System.Text;
+using System.Text.Json;
+using Microsoft.Graph.Cli.Core.Configuration;
+using Microsoft.Graph.Cli.Core.Utils;
+
+namespace Microsoft.Graph.Cli.Core.IO;
+
+public class AuthenticationCacheUtility : IAuthenticationCacheUtility {
+    private readonly IPathUtility pathUtility;
+
+    public AuthenticationCacheUtility(IPathUtility pathUtility)
+    {
+        this.pathUtility = pathUtility;
+    }
+
+    public string GetAuthenticationCacheFilePath()
+    {
+        return Path.Join(pathUtility.GetApplicationDataDirectory(), Constants.AuthenticationIdCachePath);
+    }
+
+    public async Task<AuthenticationOptions> ReadAuthenticationIdentifiersAsync(CancellationToken cancellationToken = default)
+    {
+        cancellationToken.ThrowIfCancellationRequested();
+        var path = this.GetAuthenticationCacheFilePath();
+        if (!File.Exists(path)) {
+            throw new FileNotFoundException();
+        }
+
+        using var fileStream = File.OpenRead(path);
+        var configRoot = await JsonSerializer.DeserializeAsync<Configuration.ConfigurationRoot>(fileStream, cancellationToken: cancellationToken);
+        if (configRoot?.AuthenticationOptions is null) throw new AuthenticationIdentifierException("Cannot find cached authentication identifiers.");
+        
+        return configRoot.AuthenticationOptions;
+    }
+
+    public async Task SaveAuthenticationIdentifiersAsync(string? clientId, string? tenantId, CancellationToken cancellationToken = default) {
+        cancellationToken.ThrowIfCancellationRequested();
+        var path = this.GetAuthenticationCacheFilePath();
+        var configuration = new Configuration.ConfigurationRoot {
+            AuthenticationOptions = new AuthenticationOptions {
```

Comment:
Consider using a dictionary of `AuthenticationOptions` to cover multi-tenant/multi-app scenarios since we are persisting the options on disk. The key to each element (AuthenticationOptions) can be a combo of `clientId+tenantId`, something similar to what MSAL does with [HomeAccountId](https://docs.microsoft.com/en-us/dotnet/api/microsoft.identity.client.iaccount.homeaccountid?view=azure-dotnet) in a token cache.

Reply:
I've created an issue to explore this further. Thanks Peter.